### ISSUE TITLE
4. A site-based approach

### ISSUE THREAD

**Issue by mikibonacci** [aiidateam/aiida-atomistic#4]
In this issue we comment on the description of properties in a site-based approach.

### pros:
- intuitive and straightforward definition of the properties (just a list where for each site we define the property)
- no "interference" between properties (successive kind redefinition if one property is set, may create ambiguities)
- Hubbard parameters are naturally site-based: some property is complicated to be defined via kinds.
- intersite properties (e.g. Hubbard) are easy to define in a site-based approach

### cons:
- need a mapping into kinds by means of thresholds(like magmoms), and this can be done on the fly in the plugin or as method in our StructureData ->may results in unexpected behaviors (too many starting magnetizations, i.e. too many kinds); however, we can do it in a self-consistent way (if kinds>10, increase the thresholds...)
- duplication of information with respect to kinds, which are more compact

### The `to_kinds` method

If we adopt the sites based approach, the sites-to-kinds mapping can be provided via a `to_kinds` method of the StructureData. This can support tolerances for the definition of new kinds.  
However, such procedure can be also done by hands for specific cases (for example, we may want a specific order of the kinds)

**Comment by mikibonacci** [comment]
In trying to adapt PwCalculation to use the new StructureData, I noticed a potential issue: 

(1) When we set the pseudos via the `aiida_pseudo.groups.family.pseudo.get_pseudos` method, this needs to know the kinds associated to the structure - this is still the old implementation. 
(2) Then, in the input file preparation for a PwCalculation, we need again the kinds. 

If we think about the new implementation, the kinds are obtained on the fly (not stored in the node) via the `get_kinds` method of the StructureData, which allows to define specific thresholds (or fallback into default ones) for each property. 

However, considering the specific PwCalculation case,  we should call two times the `get_kinds`, both to set the pseudos, so before the PwCalculation submission, and then after the submission (i.e. inside the PwCalculation routine for the input file generation). 
But this means that, as we need to have in both cases the same kinds, we should always define beforehand - during the StructureData instance creation - the thresholds used in the kinds definition and attach them to the StructureData: this is needed to avoid inconsistencies when we call the same methods in different places. 
But in this way, we are then defining the kinds when we generate the StructureData object, due to the fact that, given the thresholds, the kinds are uniquely determined. 

So, in the end, we are still storing information about kinds. However, now the kinds are a site-based property: for each site we have a given kind, which can be the same as the one of other sites.

In conclusion: we should always compute the `kinds` property when we generate the StructureData. Then, the user can decide to create a new StructureData with new thresholds or override explicitly the kinds (providing a corresponding input in the constructor).

**Comment by mikibonacci** [comment]
After an in person chat with @giovannipizzi, we decided that indeed the the kind should be generated when the instance is created. This can happen using thresholds (custom or default) or an explicit list of kinds. In the latter case, a check on the consistency of the value of the properties should be done, but I have some idea (I will discuss with @giovannipizzi these days). 
If the user is not happy with the generated kinds, can always generate a new structuredata with the correct ones.

**Comment by sphuber** [comment]
I am trying to remember exactly what the problems were with the kinds-centric approach of the original `StructureData` because there was some motivation to move away from it. If we were to reintroduce it, albeit as a derived property in essence, I wonder if the original problem remains unsolved.

One example that I seem to recall is that specifying magnetic moments in `PwCalculation` are necessarily kinds based, and so if one wanted to change the moments for certain sites, a new `StructureData` had to be created with new matching kinds. I think the argument was that if the structure was merely site-based, the magnetic moments could be specified independent of any kinds. But with the new design where magnetic moments are directly stored on the `structureData`, this whole discussion is moot anyway, because regardless of whether kinds are used, changing the moments would require creating a new `StructureData`. So maybe in the end the original objections to kinds are moot anyway when properties are stored on the `StructureData`. @mbercx do you recall more problems the old kinds-approach would result in?

**Comment by mikibonacci** [comment]
Additional comment on why I think is better to provide kinds in the `StructureData` and not to generate them on-the-fly (on-demand) every time:

suppose we generate on the fly the kinds, this means that we then have to do it multiple times for different plugins. For example, `aiida-pseudo` and `aiida-quantumespresso` require this kind generation.  
This means that, if the user wants to provide custom thresholds to detect kinds, he should do it for each plugin. This means that:

- we should keep track of these thresholds in the provenance
- for each plugin I need to add an input which defines these thresholds

Moreover, the kinds detection will change the value of the properties and so modify the StructureData, in such sense. But then this is out of the user control, so it may introduce indeed unwanted effects. 

I don't think this is the correct way to deal with kinds.
